File: TSVersion/Protocol/Security/SecurityHandler.ts
import { SHAuthorizeRequest, SHSecuredRequest, SHClearRequest } from "../Request"
import { SHCryptoDataResponse } from "../Response"
//обработка безопасности протокола

import * as CryptoProxy from "./CryptoProxy"


///Снимает и устанавливает защитную оболочку вокруг пакета данных
export class SecurityHandler {

    constructor(cryptoMethod: CryptoProxy.CryptoProxyMethods, cryptoProxy: CryptoProxy.ICryptoProxy) {
        this.cryptoMethod = cryptoMethod
        this.cryptoProxy = cryptoProxy
    }


    readonly cryptoMethod: CryptoProxy.CryptoProxyMethods


    readonly cryptoProxy: CryptoProxy.ICryptoProxy
    

    ///Снятие защиты
    removeShell(secured_request: SHSecuredRequest): SHClearRequest{
        let decryptedRequest = this.cryptoProxy.decrypt(secured_request.request)
        let decryptedRequestJSON = JSON.parse(decryptedRequest)

        let authorizeRequest = decryptedRequestJSON[0]
        let dataRequest = decryptedRequestJSON[1]

        let authorizeRequestTyped = new SHAuthorizeRequest()
        authorizeRequestTyped.type = authorizeRequest.type
        authorizeRequestTyped.operation = authorizeRequest.operation
        authorizeRequestTyped.user = authorizeRequest.user
        authorizeRequestTyped.password = authorizeRequest.password
        authorizeRequestTyped.time = authorizeRequest.time

        let unsecuredRequest: SHClearRequest = {
            type: secured_request.type,
            operation: secured_request.operation,
            UID: secured_request.UID,
            request: secured_request.request,
            stayAlive: secured_request.stayAlive,
            authorizeRequest: authorizeRequestTyped,
            dataRequest: dataRequest
        }

        return unsecuredRequest
    }


    ///Проверка доступа
    checkAuthorize(authorize: SHAuthorizeRequest): boolean  {
        let unixTime = new Date().valueOf()
    
        let timeCheck = (Math.abs(Number(authorize.time) - unixTime) < 5000)
        return timeCheck
    }


    putInShell(str: string, operation: string): SHCryptoDataResponse{
        let encrypted = this.cryptoProxy.encrypt(str)
        
        let cryptoResponse = new SHCryptoDataResponse()
        cryptoResponse.operation = operation
        cryptoResponse.response = encrypted
        
        return cryptoResponse
    }
}

File: untitled1/Protocol/Security/securityhandler.cpp
#include "securityhandler.h"



SecurityHandler::SecurityHandler(RSACryptoProxy *cryptoProxy, QObject *parent): QObject(parent)
{
    this->cryptoProxy = cryptoProxy;
}


QString SecurityHandler::serverKey()
{
     CryptoKey key = *(cryptoProxy->keys->self_public_key());
     return key.toString();
}


bool SecurityHandler::checkAuthorize(SHAuthorizeRequest authorize)
{
    qint64 unixTimeNow = QDateTime::currentMSecsSinceEpoch();
    qint64 unixTimeRequest = authorize.time.toUInt();
    return unixTimeNow == unixTimeRequest;
}


SHNakedRequest SecurityHandler::removeShell(SHCryptoDataRequest secured_request)
{
    QString secured_str = secured_request.request;
    QString clean_str = cryptoProxy->decrypt(secured_str);

    SHNakedRequest result;
    result.operation = secured_request.operation;
    result.request = secured_request.request;
    result.stayAlive = secured_request.stayAlive;

    SHAuthorizeRequest authorizeRequest;

    result.authorizeRequest = authorizeRequest;

    SHCryptoDataRequest dataRequest;
    dataRequest.request = clean_str;

    result.dataRequest = dataRequest;

    return result;
}


SHCryptoDataResponse SecurityHandler::putInShell(QString str, QString operation)
{
    SHCryptoDataResponse response;
    response.operation = operation;

    QString secured_Str = cryptoProxy->encrypt(str);
    response.response = secured_Str;

    return response;
}
